coverage: Treat each match arm as a "branch" for branch coverage

Zalathar opened the pull request:
This PR adds branch coverage instrumentation for individual arms in `match` expressions.

It seems to work well in most cases. However, the current implementation cannot handle match expressions containing or-patterns `|`, and therefore ignores such match expressions entirely.

(More specifically, it ignores a match expression if any of its `Candidate`s has `pre_binding_block = None`.)

---

Right now I'm not sure whether this should be merged as-is (to provide a useful subset of functionality, and avoid bit-rot), or should be left as a draft until the or-pattern limitation can be lifted (could potentially require major rework).

Zalathar committed:
coverage: Treat each match arm as a "branch" for branch coverage

Swatinem commented on compiler/rustc_mir_transform/src/coverage/mod.rs:
```diff
@@ -136,20 +136,48 @@ fn create_mappings<'tcx>(
+                false_term: rest_counter.as_term(),
+            };
+
+            mappings.extend::<Option<_>>(
+                try {
+                    let code_region = make_code_region(source_map, file_name, span, body_span)?;
+                    Mapping { kind, code_region }
+                },
+            );
```
first time I see a `try` block in action.
But isn’t this the same as:
```suggestion
            if let Some(code_region) = make_code_region(source_map, file_name, span, body_span) {
                mappings.push(Mapping { kind, code_region });
            }
```

that would look a lot more readable.

Swatinem reviewed (approved):
lgtm, the added tests really speak for themselves 🎉 
left a small style nit

Zalathar commented:
Some small style updates [(diff)](https://github.com/rust-lang/rust/compare/3f670c601be2f58554d33e493c52388969ccebc0..fa7454a856c467be5608692939c745ecff9f9064).

Zalathar commented on compiler/rustc_mir_build/src/build/coverageinfo.rs:
```diff
@@ -79,21 +84,92 @@ impl BranchInfoBuilder {
+        scope: SourceScope,
+        arms: impl Iterator<Item = (Span, Option<BasicBlock>)>,
+    ) {
+        let Some(arms) = arms
+            .map(|(span, pre_binding_block)| Some((span, pre_binding_block?)))
+            .collect::<Option<Vec<_>>>()
+        else {
+            return;
+        };
```
This is what causes us to currently not handle or-patterns, since we require every arm to have a `pre_binding_block`.

We can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.

Zalathar commented on tests/coverage/branch/match-arms.coverage:
```diff
@@ -0,0 +1,116 @@
+   LL|       |
+   LL|      3|    match value {
+   LL|      8|        Enum::D(d) if cond => consume(d),
+  ------------------
+  |  Branch (LL:9): [True: 24, False: 51]
+  |  Branch (LL:23): [True: 8, False: 16]
+  ------------------
+   LL|      4|        Enum::C(c) if cond => consume(c),
+  ------------------
+  |  Branch (LL:9): [True: 12, False: 39]
+  |  Branch (LL:23): [True: 4, False: 8]
+  ------------------
+   LL|      2|        Enum::B(b) if cond => consume(b),
+  ------------------
+  |  Branch (LL:9): [True: 6, False: 33]
+  |  Branch (LL:23): [True: 2, False: 4]
+  ------------------
+   LL|      1|        Enum::A(a) if cond => consume(a),
+  ------------------
+  |  Branch (LL:9): [True: 3, False: 30]
+  |  Branch (LL:23): [True: 1, False: 2]
+  ------------------
```
The `False` sides of these match-arm branches are higher than they should be (e.g. `False: 51` is clearly impossible).

Presumably there's some subtlety of the blocks involved that I'm not taking into account, leading to overly-high counts.

Zalathar commented:
I noticed that I accidentally included my let-else test as well, but instead of removing it, I decided to just add my if-let test as well. Both of these test show that the respective features don't support branch coverage yet.

[(diff)](https://github.com/rust-lang/rust/compare/fa7454a856c467be5608692939c745ecff9f9064..00b4b528c64458260c0c5a6721e156eee58c2b5f)

Zalathar commented:
Discard match expressions early if they have <2 arms [(diff)](https://github.com/rust-lang/rust/compare/00b4b528c64458260c0c5a6721e156eee58c2b5f..1613edf15f414055982d23807f1825a807851d5f).

Zalathar commented:
After investigating some more, I realised that my scheme for making the false-branch expressions isn't accounting for the actual control flow around match guards.

For matches with no guards, the false-count for any arm is just the sum of the execution counts of all subsequent arms.

But for an arm with a guard, we also need to subtract the number of times that the match succeeded but the guard failed. We're currently failing to do that, which is why we get inflated counts, which then mess up all earlier arms too.

Zalathar commented:
For now I've updated the implementation to skip any match expressions that have guards, so that at least we aren't producing *wrong* branch-coverage numbers for them [(diff)](https://github.com/rust-lang/rust/compare/356427f5aaf5562f3e7a6ff93c35d856a7a80b3c..b3b41a89d11dd57d4782f8dfa6c0dac89059667e).

Zalathar commented on compiler/rustc_mir_build/src/build/matches/mod.rs:
```diff
@@ -325,6 +329,30 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+        // Bail out if branch coverage is not enabled for this function.
+        let Some(branch_info) = self.coverage_branch_info.as_mut() else { return };
+
+        // FIXME(#124118) The current implementation of branch coverage for
+        // match arms can't handle or-patterns or guards.
+        if candidates.len() < 2
+            || candidates.iter().any(|c| c.pre_binding_block.is_none() || c.has_guard)
+        {
+            return;
+        }
```
This is where we currently avoid adding branch coverage for match expressions that contain any or-patterns or guards.

(This is an improvement over previous drafts, which would ICE on or-patterns, and give incorrect counts for match expressions containing guards.)

For or-patterns, we can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.

For guards, we need to track enough information to let the instrumentor subtract the guard-was-false path from the sum of subsequent arms.

Zalathar commented:
More style updates [(diff)](https://github.com/rust-lang/rust/compare/b3b41a89d11dd57d4782f8dfa6c0dac89059667e..d642bc8c69fe55ee508c8e6ee53653e6e1bd9211).

Zalathar commented on compiler/rustc_mir_build/src/build/coverageinfo.rs:
```diff
@@ -79,35 +89,91 @@ impl BranchInfoBuilder {
+            return;
+        }
+
+        // FIXME(#124118) The current implementation of branch coverage for
+        // match arms can't handle or-patterns or guards.
+        if arms.iter().any(|arm| arm.has_guard || arm.pre_binding_block.is_none()) {
+            return;
+        }
```
This is where we currently avoid adding branch coverage for match expressions that contain any or-patterns or guards.

(This is an improvement over previous drafts, which would ICE on or-patterns, and give incorrect counts for match expressions containing guards.)

For or-patterns, we can probably dig deeper into the pattern and find a suitable block, but I'll have to pay close attention to make sure that what I'm doing is correct.

For guards, we need to track enough information to let the instrumentor subtract the guard-was-false path from the sum of subsequent arms.

Zalathar commented:
It looks like the point where we know an arm's pre-binding block *and* its guard-succeeded block is just after the call to `bind_pattern` in `lower_match_arms`.

Is there a good way to forward this information to `BranchInfoBuilder`?

bors commented:
:umbrella: The latest upstream changes (presumably #124194) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
Rebasing this on top of #123409 turned out to be very difficult (due to how the branch and MC/DC code have become interdependent), so I've filed #124217 to help relieve that tension, separately from the branch coverage changes that will be built on top of it.

Zalathar committed:
coverage: Represent branches as a list of arms

Within the `InstrumentCoverage` pass, we now represent branches as a list of
arms, instead of a true/false pair, until we prepare the final table of
mappings to be attached to the MIR body.

(We then flatten the list into two-way branches by treating each arm as a
branch between its success block, and the total of all later arms.)

Currently all of the branches produced by MIR building are still two-way, but
this is a step towards allowing many-way branches.

Zalathar committed:
coverage: Represent branches as a list of arms during MIR building, too

Zalathar committed:
coverage: Add another simple test for guarded match arms

Zalathar commented:
OK, I think I've figured out guarded arms. My previous attempts were on the right track; I just needed to be *very* careful with my algebra for computing synthetic counts.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8771518866/job/24069397496) [(plain)](https://github.com/rust-lang/rust/commit/bc042ac8d981f374a265f7cb733c768e6cf6e9b3/checks/24069397496/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting rustc_next_trait_solver v0.0.0 (/checkout/compiler/rustc_next_trait_solver)
error: unresolved link to `pre_guard_marker`
   --> compiler/rustc_middle/src/mir/coverage.rs:295:31
    |
295 |     /// If this is equal to [`pre_guard_marker`], the arm has no guard.
    |                               ^^^^^^^^^^^^^^^^ no item named `pre_guard_marker` in scope
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

```

</details>

Zalathar commented:
More assertions when extracting branch arms from MIR [(diff)](https://github.com/rust-lang/rust/compare/f66890e020ba3dfe4d901d050d6d69f40a995a2f..4be7969a507563960373dfe10453113c2322e5e6).

RenjiSann commented on tests/coverage/branch/guard-simple.coverage:
```diff
@@ -0,0 +1,30 @@
+   LL|       |
+   LL|      1|fn never_taken() {
+   LL|      1|    match black_box(false) {
+   LL|      1|        _ if black_box(false) => {}
```
I am wondering why the pattern `_` and the guard `black_box(false)` are in the same code region, while there are 2 branch expressions.

I would have expected the guard to have its own counter displayed underneath
```
   LL|      1|        _ if black_box(false) => {}
                           ^1                  ^0
```

bors commented:
:umbrella: The latest upstream changes (presumably #124399) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
Brushed some of the cobwebs off this PR.

I still don't want to proceed without a plan for how to handle or-patterns, but at least this implementation is now a more up-to-date reference.

bors commented:
:umbrella: The latest upstream changes (presumably #127174) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nadrieril commented:
https://github.com/rust-lang/rust/pull/127159 should make your life easier.

Zalathar commented:
I also want to add a link here to <https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Details.20of.20match.20guard.20semantics/near/448161106>, from which I learned that the interaction between or-patterns and guards is more complex than I had thought, because the guard might need to be evaluated multiple times per match in the general case.

Dylan-DPC commented:
@Zalathar any updates on this? thanks
